### ISSUE TITLE
Naming convention

### ISSUE THREAD

**Issue by scottgonzalez** [jquery/css-chassis#4]
BEM, SUIT, OOCSS, etc.?

People seemed pretty open to what naming we use, but there also seemed to be some consensus on just picking names that make sense and being flexible instead of being tied into a specific naming system.

**Comment by battaglr** [comment]
Usually when discussing about naming conventions, we end up talking about architecture. Of course, the two are closely related and it's understandable. *Maybe we could open another issue to discuss specifically about the architecture.*

Anyhow, AFAIK —and **please** correct me if I'm wrong:

+ OOCSS doesn't have a canonical naming convention. Fortunately, as some have mention, @csswizardry has done an amazing job on this topic, but it's not that different from BEM's approach... and I like them both. But it's not an OOCSS naming convection rather a BEM variant –not like if that's a big deal.
+ SMACSS attempts to suggest a naming convention, but @snookca doesn't intent to be too prescriptive about it.
+ BEM has a more rigid naming convention, but encourages developers to adapt it to their needs.
+ SUITCSS by @necolas is the most extensively documented and maybe prescriptive naming convention so far and has a lot of great ideas –it's important to notice that it takes a lot from BEM and some from SMACSS.

IMHO we could define a new one if we feel like it, or just go with one of the previously mentioned. In my experience, a sane naming convention should consider the following:

+ Helper, utility or general purpose classes.
+ Components –blocks, or however we like to call them.
+ Sub-components –elements, or however we like to call them.
+ Modifiers.
+ States —i.e.: `.is-active`.
+ Behavior hooks —this would be how we would like to reference elements to be manipulated with JavaScript.

I know I haven't proposed any naming convention, but my intention was to take a little step back to see if we need to explicitly define what we need/want and build upon that.

Hope it helps.

**Comment by arschmitz** [comment]
@battaglr Thanks for all the information and input. As you can see we have had a TON of discussion around this, ( what you see in this issue is only the tip of the iceberg ). 

We through all this discussion, came to the decision as a team, that we wanted to go with a BEM naming convention, for a variety of reasons. The core of these reasons came down to, performance, flexibility, and ease of use, for the cases we are trying to address. One thing I would like to mention, that I think is a common misconception, is that BEM  has a super strict naming convention. This was something I had been wrong about too, until we talked with @veged and others from Yandex ( creators of BEM ).

The main thing we are still trying to decide here right now, is exactly what flavor of BEM, we would like to go with. ( since as i just mentioned they are not at all strict about your exact class name structure ). The whole thing with the dashes double dashes underscores ect, is just the most common "flavor" of BEM.

**Comment by battaglr** [comment]
@arschmitz: I agree, and believe that choosing some flavor of BEM is right decision. If I find the time I'll try to help shaping this new "flavor".

**Comment by necolas** [comment]
FWIW, SUITCSS doesn't take anything from SMACSS. BEM-like conventions are about little more than scoping CSS to the DOM tree of a component. In a way, I think inline-styles are probably a better solution.

**Comment by battaglr** [comment]
@necolas: SMACSS didn't first documented the `.is-something` pattern? That comment was meant to give some sort of context and nothing else, but I'm glad that you corrected me on that. :)

Can you please elaborate about the BEM-like conventions? If I get what you're saying, I agree that in some cases or in the original methodology you try to scope and describe the whole DOM tree (`.something`, `.something__other`, `.something__other__item`, `.something__other__item__yet-another-stuff`, instead of sometimes do `.something__yet-another-stuff`), and that is not a good idea if you want your code to be flexible. If is that, we agree.

And I may be missing way too much, but alway saw SUITCSS naming convention as something very similar to BEM —with a lots of improvements. Please, correct me if I didn't get that correctly.

**Comment by necolas** [comment]
Yeah it's very similar, with a bit of tooling around it. BEM-like patterns are for component-based designs; ways of styling the known DOM tree of a component, scoping those styles somewhat. Inline styles also work for that (better in many ways).

**Comment by battaglr** [comment]
@necolas: thanks for the explanation! You mentioned inline styles —I'm thinking on `style="property: value;"`— again (and this may be off topic but I'm sure it will enrich the discussion), but I'm having a hard time trying to get why you're saying that inline styles may work better; I always saw inline styles as something to avoid if you could, but as I respect your opinion very much, I'm wondering if you could you elaborate on that, please.
